truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -264,78 +264,159 @@ public Object ulong2num(long num) {
+    public abstract static class IntegerBytesNode extends CoreMethodArrayArgumentsNode {
+
+        // The Ruby MRI API for extracting the contents of a integer
+        // fills a provided buffer of words with a representation of
+        // the number in the specified format. This allows the users
+        // to specify the word order in the buffer, the endianness
+        // within the word, and whether to encode the number as a
+        // magnitude or in two's complement.
         //
-        // Java BigInteger provides an API to extract the two's
-        // complement as a byte array starting with the most
-        // significant byte and continuing until the number is
-        // complete. To convert from this to the Ruby form we must:
+        // The API also returns an integer indicating whether the
+        // number was positive or negative and the overflow
+        // status. That part is implemented in C while this Java
+        // method is purely concerned with extracting the bytes into
+        // the buffer and getting them in the right order.
         //
-        // 1. Starting from the end of the byte array pack 8 bytes
-        // into a long (being careful to avoid accidental sign
-        // extension occurring due to widening of primitive types).
+        // If the buffer is too short to hold the entire number then
+        // it will be filled with the least significant bytes of the
+        // number.
         //
-        // 2. Pack any remaining bytes into a final long and pad it
-        // appropriately to preserve the sign of the number.
+        // If the buffer is longer than is required for encoding the
+        // number then the remainder of the buffer must not change the
+        // interpretation of the number. I.e. if we are encoding in
+        // two's complement then it must be filled with 0xff to
+        // preserve the number's sign.
         //
-        // 3. Fill any remaining space in the requested buffer size
-        // with longs (again preserving the sign of the number).
-        //
-        // It is important we correctly fill buffers which are longer
-        // than requested both for compatibility with MRI and to
-        // preserve the returned number.
+        // The API allows the order of words in the buffer to be
+        // specified as well as the order of bytes within those
+        // words. We separate the process into two stages to make the
+        // code easier to understand, first copying the bytes and
+        // adding padding at the start of end, and then reordering the
+        // bytes within each word if required.
+
+        @Specialization
+        @TruffleBoundary
+        public DynamicObject bytes(int num, int num_words, int word_length, boolean msw_first, boolean twosComp, boolean bigEndian) {
+            BigInteger bi = BigInteger.valueOf(num);
+            return bytes(bi, num_words, word_length, msw_first, twosComp, bigEndian);
+        }
```

Comment:
You don't need an explicit `int` specialization as it is implicitly tried to cast it as a `long` by the DSL:
https://github.com/graalvm/truffleruby/blob/fab462f11faedf2bf8e5ed0a653c94988db02776/truffleruby/src/main/java/org/truffleruby/language/RubyTypes.java#L42-L45